samza-core/src/test/java/org/apache/samza/processor/TestStreamProcessor.java

```diff
@@ -165,10 +175,98 @@ public void onFailure(Throwable t) {
     }
   }
 
+  /**
+   * Tests that a failure in container correctly stops a running JobCoordinator and propagates the exception
+   * through the StreamProcessor
+   *
+   * Assertions:
+   * - JobCoordinator has been stopped from the JobCoordinatorListener callback
+   * - StreamProcessorLifecycleListener#onFailure(Throwable) has been invoked
+   */
+  @Test
+  public void testContainerFailureCorrectlyStopsProcessor() {
+    JobCoordinator mockJobCoordinator = mock(JobCoordinator.class);
+    Throwable expectedThrowable =  new SamzaException("Failure in Container!");
+    AtomicReference<Throwable> actualThrowable = new AtomicReference<>();
+    final CountDownLatch runLoopStartedLatch = new CountDownLatch(1);
+    RunLoop failingRunLoop = mock(RunLoop.class);
+    doAnswer(invocation ->
+      {
+        try {
+          Thread.sleep(10);
+          runLoopStartedLatch.countDown();
+          throw expectedThrowable;
+        } catch (InterruptedException ie) {
+          ie.printStackTrace();
+        }
+        return null;
+      }).when(failingRunLoop).run();
+
+    SamzaContainer mockContainer = StreamProcessorTestUtils.getDummyContainer(failingRunLoop, mock(StreamTask.class));
+
+    final CountDownLatch processorListenerFailed = new CountDownLatch(1);
+
+    TestableStreamProcessor processor = new TestableStreamProcessor(
+        new MapConfig(),
+        new HashMap<>(),
+        mock(StreamTaskFactory.class),
+        new StreamProcessorLifecycleListener() {
+          @Override
+          public void onStart() { }
+
+          @Override
+          public void onShutdown() { }
```

Comment:
It would be nice to verify that shutdown is not also called when the processor fails due to a user exception.